src/js/addons/widgets/ColorPicker.js

```diff
@@ -2,34 +2,48 @@
 /* global thistle */
 
 import Widget from './Widget.js';
-import { toCSS, getPosition } from '../../core/common.js';
+import { toCSS } from '../../core/common.js';
 
 export default class ColorPicker extends Widget {
-    constructor (datum) {
-        super(datum);
+    // There must be a constructor call here because
+    // ColorPicker sets its own properties for color
+    constructor (...args) {
+        super(...args);
+
+        // Set the color property from editor value
+        this.color = toCSS(this.key.value);
     }
 
-    create (keyPair, cm) {
+    createEl () {
         let el = document.createElement('div');
+
         el.className = 'tp-widget tp-widget-colorpicker';
-        el.value = String(keyPair.pos.line) + '-' + String(keyPair.index);
-        el.style.background = toCSS(keyPair.value);
 
-        el.addEventListener('click', function (e) {
-            let picker = new thistle.Picker(el.style.background);
-            let pos = getPosition(el);
+        el.addEventListener('click', (e) => {
+            let picker = new thistle.Picker(this.color);
+            let pos = this.getPosition();
 
-            picker.presentModal(pos.x + 20,
-                                cm.heightAtLine(parseInt(el.value)) + 20);
+            picker.presentModal(pos.x + 20, pos.y + 20);
+            picker.on('changed', (event) => {
+                this.color = picker.getCSS();
 
-            picker.on('changed', function () {
-                el.style.background = picker.getCSS();
+                // Convert the CSS color value to Tangram format for editor.
                 let color = picker.getRGB();
-                let str = '[' + color.r.toFixed(3) + ',' + color.g.toFixed(3) + ',' + color.b.toFixed(3) + ']';
-                cm.tangram_play.setValue(cm.tangram_play.getKeyForStr(el.value), str);
+                let rgbString = '[' + color.r.toFixed(3) + ',' + color.g.toFixed(3) + ',' + color.b.toFixed(3) + ']';
+                this.setEditorValue(rgbString);
             });
         });
 
-        return this.wrap(el, keyPair);
+        return el;
+    }
+
+    get color () {
```

Comment:
[Getters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) and [setters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set) have been part of ES5 and have been supported in browsers for a very long time. I started using them with ES6 though because I really like their syntax here. This lets you set a "color" property on the object but whenever you do that, it also changes the widget's background color in CSS, so that you reduce the chance that you might change a value but forget to change a background color (and hence the values are out of sync). It's not perfect here because there's a difference between a CSS background color and the RGB string that Tangram prefers, so there's definitely room for improvement.